rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+nuance between well-optimized program-local synchronization and worst-case FFI
+synchronization. See Unresolved Questions section for more details.
+
+However, the fact that these operations do not necessarily compile into a single
+hardware instruction is arguably a footgun for volatile's use cases, and it
+could be argued that initially only stabilizing loads, stores and `Relaxed`
+atomic ordering would be more prudent. This will be revisited in the
+alternatives section of this RFC.
+
+---
+
+As currently designed, this RFC uses `arbitrary_safe_types` to give method-like
+semantics to a `NonNull` raw pointer. This seems necessary to get reasonable
+ergonomics with an atomics-like wrapper type approach. However, it could also be
+argued that a `VolatileXyz::store(ptr, data, ordering)` style of API would work
+well enough, and avoid coupling with unstable features. Similarly, the use of
+`NonNull` itself could be debated. This will be revisited in the alternatives
+section of the RFC as well.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Deprecating the existing volatile operations will cause language churn. Not
+deprecating them will keep around two different and subtly incompatible ways to
+do the same thing, which is equally bad. It could be argued that the issues with
+the existing volatile operations, while real, do not warrant the full complexity
+of this RFC.
+
+As mentioned above, exposing more than loads and stores, and non-`Relaxed`
+atomic memory orderings, also muddies the "a volatile op should compile into one
+hardware memory instruction" narrative that is so convenient for loads and
+stores. Further, compatibility of non-load/store atomics in IPC scenario may
+require some ABI agreement on how atomics should be implemented between the
+interacting programs.
+
+If this is perceived to be a problem, we could decide to do away with some of
+the complexity by initially focusing on a restricted subset of this proposal
+which only supports `Relaxed` loads and stores, and saving more complex atomic
+operations as a future extension.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This effort seems worthwhile because it simultaneously eliminates two well-known
+and unnecessary volatile footguns (data races and tearing) and opens interesting
+new possibilities in interprocess communication.
+
+But although this feature may seem simple, its design space is actually
+remarquably large, and a great many alternatives were considered before reaching
+the current design proposal. Here are some design knobs that were explored:
+
+## Extending `AtomicXyz`
+[extending-atomicxyz]: #extending-atomicxyz
+
+Atomic volatile operations could be made part of the Atomic wrapper types.
+However, they cannot be just another atomic memory ordering, due to the fact
+that atomic ops take an `&self` (which is against the spirit of volatile as it
+is tagged with `dereferenceable` at the LLVM layer).
```

Comment:
I think this is a red herring; `&self` is (almost) as bad for `AtomicX` as it is for `VolatileX`. We'll have to resolve this issue anyway for `AtomicX`. If we manage to get rid of `dereferencable` for the `AtomicX` methods, then we can make "volatile" an `Ordering`. (We might also want to just weaken the attribute to something like dereferencable-on-entry-to-the-function, which would *not* be good enough for volatile operations. But I think unifying volatile and atomic accesses here is a good motivation for doing the former.)

Reply:
Let's try to collect all discussion of `&self` in `AtomicX` and other `UnsafeCell`s on the thread above, for now, unless you feel that some part of it belongs here.

I _dislike_ the idea of making `Volatile` an atomic `Ordering`, because since `Ordering` does not have set semantics (which in my opinion is also a mistake, but I know you disagree and I do not have enough time to litigate that particular issue), it forever sets in stone the fact that volatile and atomic are not orthogonal and non-`Relaxed` atomic volatile will never be a thing.

OTOH, it seems @gnzlbg and I won't manage to convince others that pessimistically assuming the continued existence of arbitrary FFI threads which may read and write to every pub static variable and every pointer that was ever sent to FFI at some point in program execution (or _may_ have been sent to FFI, in absence of inlining) is not a smart idea and will harm the optimization of mostly-Rust codebases. If that is the case, then I guess making atomic an effective superset of volatile for the purpose of shared-memory IPC is not an issue.

Reply:
> I dislike the idea of making Volatile an atomic Ordering, because since Ordering does not have set semantics (which in my opinion is also a mistake, but I know you disagree and I do not have enough time to litigate that particular issue with no clear benefit in sight), it forever sets in stone the fact that volatile and atomic are not orthogonal 

Adding it to `Ordering` is a hack. Clearly the right thing to do is to add a new `volatile: bool` (or rather a newtype-enum) to each operation to properly express that they are orthogonal. Alas, that is not API-compatible. Hence the proposal to use an `Ordering`.

> and non-Relaxed atomic volatile will never be a thing.

When I said "add to `Ordering`" I totally had in mind that in the long run, we would have `Ordering::RelaxedVolatile`, `Ordering::SeqCstVolatile`, and so on. So that they are definitely a thing! Sorry for not being clearer.

> If that is the case, then I guess making atomic an effective superset of volatile for the purpose of shared-memory IPC is not an issue, because all the optimization potential that emerges from distinguishing between non-volatile atomic (normal everyday synchronization within well-understood Rust code) and volatile atomic (weird edge case synchronization which compiler does not have a complete picture of) is effectively already lost.

While I am not a fan of the strong "Rust isolation" property you are proposing, I also do not understand what it has to do with the discussion we are having here that is purely about how the exposed API looks like, not its semantics. I view those as largely orthogonal concerns here.

Reply:
Thanks for the clarifications.

Just to be clear, how does your proposal of having `Ordering::RelaxedVolatile` here interact with your other proposal below to have volatile reads and writes use `Unordered` LLVM ordering for maximal hardware portability? Is it really okay to expose an `Ordering::UnorderedVolatile` even though we're not really ready to commit to exposing `Unordered` as an atomic ordering in Rust?

Reply:
>  Is it really okay to expose an Ordering::UnorderedVolatile even though we're not really ready to commit to exposing Unordered as an atomic ordering in Rust?

No.

My proposal was to have one API for "non-atomic" use that is intended to supersede the currently stable API, with the goal of (a) fixing implicit tearing, and while we are at it also (b) clarify the semantics of races. That one would internally use "unordered volaltile", but on the outside it would just be "volatile in a way that races are not UB (nor poison)". This would provide just "load" and "store" that each lower to a single instruction. It would be the recommended API for MMIO.

My impression was that "atomic volatile" is something of a dedicated use-case that's a bit further out there, and we don't have to handle everything with one API.

TBH these feel like two related but separate RFCs to me. The first is about fixing the volatile we currently have, the second is about expanding what can be done with volatile (things like volatile RMWs or `SeqCstVolatile`).

Reply:
Thanks for clarifying. Also, I noticed that one of your points above called for a reply, so...

> While I am not a fan of the strong "Rust isolation" property you are proposing, I also do not understand what it has to do with the discussion we are having here that is purely about how the exposed API looks like, not its semantics. I view those as largely orthogonal concerns here.

[Earlier in this thread](https://github.com/rust-lang/unsafe-code-guidelines/pull/212#issuecomment-543262839), @Amanieu stated that nontrivial atomic volatile semantics (i.e. more than `Unordered` loads and stores) were useless in his view because...

1. MMIO needed nothing more than unordered volatile loads and stores, or more specifically it required unordered loads and stores together with special hardware memory barriers that should fall outside of the scope of any sane language memory model
2. Shared memory IPC needed nothing more than regular atomics without volatile qualifiers.

Point 2 was based on a certain vision of FFI where any memory location which ever interacted with FFI (be it by being a public symbol of the binary or by being associated with a pointer which may ever have been received from or sent to an FFI function) must be effectively assumed to be continuously accessed by adversarial FFI threads throughout the rest of the execution of the program, except where the "data races are UB" rule can be used to rule out this behavior.

So far, this vision seems to the winning one. If it is the memory model that we're going for, then indeed we don't need any more volatile atomics semantics than provided by unordered loads and stores.

Reply:
Okay, correcting myself in the subset of the thread below (ping @comex), I think I just found what use nontrivial forms of atomic volatile could actually have.

At this point in time, LLVM (and by extension rustc) treats atomics with extreme caution, [far beyond what C++11 semantics actually mandate](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4455.html). Some performance-conscious people would like this to change, while [others are scared about the implications](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0062r1.html).

One possible way to resolve this social tension would be to say that non-volatile atomics are freely optimizable by the compiler, and special constructs should be used where people want to exercise more precise control on which atomic operations get actually executed. Fences are sometimes mentioned as one such construct, but atomic volatile could be an even stronger alternative in crypto-like cases where people effectively want to use Rust as a high-level assembler.

Reply:
Integrated.

Reply:
I could also totally imagine an MMIO situation where a concurrent data structure (say, a concurrent ring buffer) is shared between the CPU and an external device, and atomic MMIO is needed for coordination between the two. Or is that not needed because CPUs treat MMIO specially and the caching effects (that usually mandate synchronization and fences even on the CPU level) don't show up there?